Sources/Algorithms/Reductions.swift

```diff
@@ -0,0 +1,308 @@
+      return result
+    }
+  }
+
+  public func makeIterator() -> Iterator {
+    Iterator(iterator: base.makeIterator(),
+             current: initial,
+             transform: transform)
+  }
+}
+
+extension ExclusiveReductions: Collection where Base: Collection {
+  public typealias Index = ReductionsIndex<Base.Index, Result>
+
+  public var startIndex: Index { Index(representation: .start) }
+  public var endIndex: Index { Index(representation: .end) }
+
+  public subscript(position: Index) -> Result {
+    switch position.representation {
+    case .start: return initial
+    case .base(_, let result): return result
+    case .end: fatalError("Cannot get element of end index.")
+    }
+  }
+
+  public func index(after i: Index) -> Index {
+    func index(base index: Base.Index, previous: Result) -> Index {
+      guard index != base.endIndex else { return endIndex }
+      return .base(index: index, result: transform(previous, base[index]))
+    }
+    switch i.representation {
+    case .start:
+      return index(base: base.startIndex, previous: initial)
+    case let .base(i, result):
+      return index(base: base.index(after: i), previous: result)
+    case .end:
+      fatalError("Cannot get index after end index.")
+    }
+  }
+
+  public func distance(from start: Index, to end: Index) -> Int {
+    switch (start.representation, end.representation) {
+    case (.start, .start):
+      return 0
+    case let (.start, .base(index, _)):
+      return base.distance(from: base.startIndex, to: index) + 1
+    case (.start, .end):
+      return base.distance(from: base.startIndex, to: base.endIndex) + 1
+    case let (.base(index, _), .start):
+      return base.distance(from: index, to: base.startIndex) - 1
+    case let (.base(start, _), .base(end, _)):
+      return base.distance(from: start, to: end)
+    case let (.base(index, _), .end):
+      return base.distance(from: index, to: base.endIndex)
+    case (.end, .start):
+      return base.distance(from: base.endIndex, to: base.startIndex) - 1
+    case let (.end, .base(index, _)):
+      return base.distance(from: base.endIndex, to: index)
+    case (.end, .end):
+      return 0
+    }
+  }
+}
+
+extension ExclusiveReductions: LazySequenceProtocol
+  where Base: LazySequenceProtocol {}
+
+extension ExclusiveReductions: LazyCollectionProtocol
+  where Base: LazyCollectionProtocol {}
+
+// MARK: - Inclusive Reductions
+
+extension LazySequenceProtocol {
+
+  public func reductions(
+    _ transform: @escaping (Element, Element) -> Element
+  ) -> InclusiveReductions<Self> {
+    InclusiveReductions(base: self, transform: transform)
+  }
+}
+
+extension Sequence {
+  public func reductions(
+    _ transform: (Element, Element) throws -> Element
+  ) rethrows -> [Element] {
+    var iterator = makeIterator()
+    guard let initial = iterator.next() else { return [] }
+    return try IteratorSequence(iterator).reductions(initial, transform)
+  }
+}
```

Comment:
These are missing documentation as well. Might also be worth pointing out the relation to the version of `reductions` that takes an initial result.